### ISSUE TITLE
FEATURE REQUEST: New variable 'pane_current_command_with_args'

### ISSUE THREAD

**Issue by jamespharvey20** [tmux/tmux#1533]
Current variable `pane_current_command` only shows the process name, not the arguments.

I'd like a new variable that shows the arguments as well.  My intention would be in the `choose-tree` format string, to drop `${pane_title}` and use this instead of `${pane_current_command}`.  (I know I'd have to deal with the case of a window line with only 1 pane as well.)

This is my attempt to be able to find processes better that don't set a terminal title.  Given this variable, I'd probably experiment with adding to the format string to detect if `${pane_title}` was set/nonempty, and if so, using it for things that support it.  But then modifying my default environment (`.bashrc`?) so it is an empty string by default.  So, I'd probably only have `choose-window` show a command with arguments as a fallback to a meaningful terminal title.

Really long commands could be so long they'd be larger than 1 line.  Making `choose-tree` very small now shows it truncates the line, which I'm fine with, so hopefully this wouldn't cause a problem even with really long arguments.

**Comment by jamespharvey20** [comment]
I tried implementing this, for `linux` only, really thought I had it, but it didn't affect the output.

`pane_current_command()` only shows the process name because `format_cb_current_command()` calls `osdep_get_name()`, which includes:

    while (( ch = fgetc(f)) != EOF) {
       if ( ch == '\0')
          break
    ...

This is parsing `/proc/<pid>/cmdline`, which separates the process name and each of the arguments with `\0`.  Use `cat -e /proc/<pid>/cmdline` to see this.

**Patch: http://termbin.com/r9va**

In `osdep-linux.c`, I duplicated `osdep_get_name()` to `osdep_get_name_with_args()`, and changed this to:

    while (( ch = fgetc(f)) != EOF) {
       if ( ch == '\0')
          ch = ' '
    ...

By extracting the function and running it manually, I see this change works as expected, and my own program calling it correctly prints the process name with all of the arguments.

**Patch: http://termbin.com/yfgs**

I then duplicated `format_cb_current_command()` to `format_cb_current_command_with_args()`, instead calling `osdep_get_name_with_args()`, and included the necessary `format_add_cb()` "registration" of the new command.

Yet, when using `${pane_current_command_with_args}` with `choose-tree -F`, it doesn't change whatsoever.  It still shows just the process name.

**Comment by jamespharvey20** [comment]
I hope this sounds worthwhile to add.  I realize with all the osdep variants, this could be a bit of work & testing.  :-(

If it's not worthwhile to add, maybe someone can give a suggestion why it's still just showing the process name, if they have one.

**Comment by nicm** [comment]
I don't want this, you will need to use a script. I don't remember if the
tree format expands #(), if not then you will need to run it some other way
and put the result in a user option.

**Comment by jamespharvey20** [comment]
> I don't want this, you will need to use a script. I don't remember if the tree format expands #(), if not then you will need to run it some other way and put the result in a user option.

Understood.

I figured out why my earlier `#{current_command_with_args}` patch was omitting the args.  I missed that `parse_window_name()` clobbers all the args.  Changing that to `xstrdup(cmd)` makes it work as intended, showing all the arguments.  But, I'd prefer to not need to patch every upgrade, and respect that variable is not wanted, so I'd like to find another way...

`#()` does expand here.  Such as:

    #(cat /proc/#{pane_pid}/cmdline | sed "s|\x0| |g')

But, I'm running into 2 problems.

1. I don't see how to reliably get the current command's pid.  `#{pane_pid}` is for the "PID of the first process in pane", which at least in all of my cases is just `bash`.

Could `#{pane_current_command_pid}` be added?

Working patch for it here: http://termbin.com/g5es

The only other way I can find to maybe get that pid is to use `#{pane_pid}`, which is for the shell, then use something like `ps -o pid --no-headers --ppid <shell pid>`, but when there's multiple child processes, I'm not sure how an external script would know which to pick as the current one.

2. Even using the pid patch, since `#()` doesn't wait, it starts with a ton of:

`...<'cat /proc/<pid>/cmdline | sed "s|\x0| |g"' not ready>`

(and, I guess it's possible if a pid recycles an old cached value might be shown)

Could a variant of `#()` be added that does wait?  With the understanding that a slow command will cause lag?

Or, maybe there's a way for a script to force those to load before it runs `choose-tree`...

Because of this problem, I'm thinking user options would be better.  But, I haven't used them before, and am not understanding from the documentation how those would help.  If instead of running `choose-tree` directly, I ran a script that first set something like `@<pane_pid>_cmdline` for each pane, then that script launched `choose-tree`, if the format string could somehow use that, I'd be all set.  But trying `@#{pane_id}_cmdline` in the format string to access it doesn't work.  I wasn't thinking it would, since there is no way to parse it easily, like `{` `}` wrapping.  I don't see how on a per-line basis it would be able to use a different user option.  Actually, I'm not seeing at all how a format string would use a user option even with a fixed name.  Including `@test` or `@{test}` in a format string, after having run `tmux setw -q @test "hi"`, doesn't use it.

**Comment by nicm** [comment]
You can use pane_tty and then use ps or /proc to work out the processes
running in the pane.

**Comment by tomskopek** [comment]
Hey @jamespharvey20,
Did you ever get anywhere with this? I would find it very useful. For example, if I have a rails server and a rails console running, they both just say "ruby".
Thanks!

**Comment by jamespharvey20** [comment]
@tomskopek, yes, but only by running a patch on top of git commit g849d06c that adds a patch from ThomasAdam for allowing choose-* command format strings to change color, and code I wrote to implement on Linux only: `pane_current_command_with_args` and `pane_current_command_args`.  Works great for me.

I just tried applying the patch on top of current git master, and it doesn't apply cleanly anymore.  ThomasAdam's commit was soon added after the commit I'm using, but it's other changes that are the real problem.  I'll need to update it at some point, but can't at the moment, so it would need to be applied on this commit or modified.  https://pastebin.com/raw/wAifLh9C

I believe I tried using `pane_tty` and using `ps` or `/proc`, but can't remember what problems I ran into.  I think it was the same problem of `$()` not waiting, the first load didn't work, and subsequent ones showed data from the last run.

I also run with this change, to skip printing the name of the tree node, and leave it up to the user formattable string, so it can be color customized.  I color attached sessions blue, unattached ones red.  I left that out of the above linked patch, as you might not want this part:

```
diff --git a/mode-tree.c b/mode-tree.c
index eb18bdf7..dd18df67 100644
--- a/mode-tree.c
+++ b/mode-tree.c
@@ -572,8 +572,8 @@ mode_tree_draw(struct mode_tree_data *mtd)
                        tag = "*";
                else
                        tag = "";
-               xasprintf(&text, "%-*s%s%s%s: %s", keylen, key, start,
-                   mti->name, tag, mti->text);
+               xasprintf(&text, "%-*s%s%s%s", keylen, key, start,
+                   tag, mti->text);
                free(start);
 
                if (mti->tagged) {
```

I then write `tmux-choose-tree` which uses these to make (to me what is) a really clean colorized choose-tree: https://pastebin.com/raw/YxJDjnTR

Ignore the red row near the top, that's from running through a terminal called Terminator.

![Color Choose Tree](https://i.imgur.com/g404xJd.png)

**Comment by jamespharvey20** [comment]
@tomskopek, I finally got around to updating this.  This is a patch that applies cleanly on current master, bc72cf2f.  https://pastebin.com/raw/sbJnhBhD

I also modified my `tmux-choose-tree` script to add a dark gray `pane_tty`, which helps me locate something through `ps`, but doesn't get in the way when I'm not looking for it.  https://pastebin.com/raw/WYdrfVsW

**Comment by tomskopek** [comment]
Hey @jamespharvey20,

Thanks a lot!

I manually changed the code according to your pastebin (btw, is there a nicer way to apply those changes using that format? I did it manually but there's gotta be a better way)

Next I tried to `./configure && make`, and I got an error:

```
Undefined symbols for architecture x86_64:
  "_osdep_get_name_args", referenced from:
      _format_cb_current_command_args in format.o
  "_osdep_get_name_with_args", referenced from:
      _format_cb_current_command_with_args in format.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I think it's because I'm on MacOSX, instead of linux. If I understand correctly, I would need to make changes to `osdep-darwin` instead of `osdep-linux`. And judging from the difference between the existing methods in those two files, I'm guessing it probably means a completely different way of fetching the arguments?

To be honest, it's a bit over my head. But thanks all the same for your generosity and for sharing!

Best,
Tom
